Database Tests

andersonjackc opened the pull request:
On top of Database tests, many IsEquals were added.

The tests are all for Get/Check queries, as we need to discuss the possible implications of modifying the database in tests.  Currently I dont see a good way of setting up a 
if(noUsersConnected){ runTests}
type check, as there are always connected threads.  On top of that, if this were to go live, the tests would theoretically never be runnable, making them unnecessary in the first place...either way, now that the IsEquals() functions are implemented, it will make for much easier test writing if we do decide to test them.

andersonjackc committed:
Added IsEqual for Admins

andersonjackc committed:
Added IsEqual for AnswerSubmissions

andersonjackc committed:
Added IsEqual for Problems

andersonjackc committed:
Added 2 different StandardUser IsEquals

andersonjackc committed:
Added IsEqual for Submissions

andersonjackc committed:
Added a Special Init for DB Tests

After talking with Josh, this can be removed, and the original Init() can be changed, once info.txt is added to the gitignore.  This may lead to AWS hosting....

andersonjackc committed:
Create DBTests.cs

Tests for every Get/Check except for 2 "impossible" to check queries.

TBD whether or not we will test Inserts/Updates/Deletes

jgross11 commented:
`SHOW STATUS LIKE 'Threads_Connected';` returns 3 when server not running, 4 when it is - could we not define a "DB_THREADS_IDLE_COUNT" or some such constant to represent the count when the server isn't running and check in the init for tests method whether the value returned from the db is equal to that before allowing tests to run?

jgross11 commented on Project EFT/Data Classes/Admin.cs:
```diff
@@ -28,5 +28,10 @@ public static Admin Decode()
         {
             return new Admin();
         }
+
+        public bool IsEqual(Admin otherAdmin)
+        {
+            return (this.Username.Equals(otherAdmin.Username) && this.Password.Equals(otherAdmin.Password) && this.Email.Equals(otherAdmin.Email) && this.Id == otherAdmin.Id);
+        }
```
no check for submission equality?

andersonjackc replied:
The test this is used for, ```testAdminLogin()``` returns an Admin with no submission list.  I have added an ```IsEqualWithSubList(Admin otherAdmin)``` in my next commit now, but it is currently not used.  Perhaps it would be used in some form of insert/delete?

jgross11 commented on Project EFT/Data Classes/StandardUser.cs:
```diff
@@ -43,5 +43,45 @@ public void UpdateRanking()
+                if (this.Submissions.ContainsKey(kv.Key))
+                {
+                    int count = 0;
+                    foreach (AnswerSubmission a in kv.Value)
+                    {
+                        if (!this.Submissions[kv.Key][count].IsEqual(a))
+                        {
+                            subMapEqual = false;
+                            shouldBreak = true;
+                        }
+                        count++;
```
sub map not equal --> users not equal, so it could just return false inside of the conditional on 61, and if the conditional never evals to true, it's implied that the submissions are equal

andersonjackc replied:
Yep, thats what my logic was doing but more convoluted.   I just wasnt thinking about the easier way to do that.  Updated.

jgross11 commented on Project EFT/Data Classes/StandardUser.cs:
```diff
@@ -43,5 +43,45 @@ public void UpdateRanking()
+                    subMapEqual = false;
+                    break;
+                }
+            }
+            return (this.Username.Equals(otherUser.Username) && this.Password.Equals(otherUser.Password) && this.Email.Equals(otherUser.Email) && this.Id == otherUser.Id && subMapEqual);
+        }
```
paired with the above comment, this could be simplified to just return the result of `IsEqual` as execution would only reach this point if the submissions are known to be equal

andersonjackc replied:
Wouldn't that add an extra function call, making it more complex memory wise?
For now I left it the way it is, minus the ```&&  subMapEqual``` boolean...

jgross11 commented on Project EFT/Data Classes/AnswerSubmission.cs:
```diff
@@ -16,5 +17,15 @@ public AnswerSubmission(string content, DateTime submissionDate, int id, bool is
+
+        //for posterity sake, this used to be an override of Equals, which required a custom GetHashCode function as well...it was never committed, for good reason...
+        //I will allow you to picture what that monstrosity looked like
+        public bool IsEqual(AnswerSubmission subs)
+        {
+            return (this.Content.Equals(subs.Content)
+                    && this.SubmissionDate.Equals(subs.SubmissionDate)
+                    && this.UserID == subs.UserID && this.IsCorrect == subs.IsCorrect
+                    && this.ProblemId == subs.ProblemId);
+        }
```
should probably override a virtual method in parent class to disallow comparing AnswerSubmission with Submission, unless that is desired

andersonjackc replied:
I tried making the Submission classes ```IsEqual(Submission subs)``` virtual, but then since the AnswerSubmission's IsEqual is defined as ```IsEqual(AnswerSubmission subs)``` it does not override the method in the parent class, since they have different parameters input.   https://stackoverflow.com/questions/27547122/c-sharp-override-with-different-parameters

I left it as is for now.

jgross11 commented on Project EFT/Data Classes/Submission.cs:
```diff
@@ -17,5 +18,15 @@ public Submission(string content, DateTime submissionDate, int userID)
+
+        //for posterity sake, this used to be an override of Equals, which required a custom GetHashCode function as well...it was never committed, for good reason...
+        //I will allow you to picture what that monstrosity looked like
+        public bool IsEqual(Submission subs)
+        {
+            return (this.Content.Equals(subs.Content)
+                    && this.SubmissionDate.Equals(subs.SubmissionDate)
+                    && this.UserID == subs.UserID);
+        }
+
```
should probably be virtual to allow subclasses to override and to avoid subclasses being able to check if they're equal to base / other subclasses besides their own

andersonjackc replied:
Addressed in above reply.  When this is marked as virtual, if the AnswerSubmissions is overridden, then the input parameter would have to be Submission, not AnswerSubmission, requiring a cast with a try/catch

jgross11 commented on Project EFT/Database/DBConnector.cs:
```diff
@@ -33,6 +33,18 @@ public static void Init()
             if (!GetNumberUsersInDB()) throw new Exception("Could not get initial number of users in DB");
         }
 
+        public static void InitForTests()
+        {
+            string[] lines = System.IO.File.ReadAllLines("../../../../../info.txt");
+            connectionString = @"server=" + lines[0] + ";userid=" + lines[1] + ";password=" + lines[2] + ";database=" + lines[3];
+
+            // cache list of all problems currently in db
+            if (!GetProblemsList()) throw new Exception("Could not read initial problem list from DB");
+
+            // cache number of users in db
+            if (!GetNumberUsersInDB()) throw new Exception("Could not get initial number of users in DB");
+        }
+
```
not sure if the gitignore solution would fix this entirely - I can see the case where the actual init no longer needs "../../" before info.txt, but the test init still needs the first three, though that's just a guess

jgross11 commented on Project-EFT Tests/DBTests.cs:
```diff
@@ -0,0 +1,178 @@
+            List<AnswerSubmission> testResults = DBConnector.GetAnswerSubmissionsByID(userID);
+
+            //iterates through the specific list and checks to see if they equal the expected values set in the setup function
+            int counter = 0;
+            foreach(AnswerSubmission sub in testResults)
+            {
+                Assert.True(sub.IsEqual(usersSubmissions[counter]));
+                counter++;
+
+                if (counter > usersSubmissions.Count)
+                {
+                    break;
+                }
+            }
```
should there be a test to ensure the returned list has the expected count?

andersonjackc replied:
Updated it now.

jgross11 commented on Project-EFT Tests/DBTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+            //iterates through the specific list and checks to see if they equal the expected values set in the setup function
+            int counter = 0;
+            foreach (Submission sub in testResults)
+            {
+                Assert.True(sub.IsEqual(adminSubmissions[counter]));
+                counter++;
+
+                if (counter > adminSubmissions.Count)
+                {
+                    break;
+                }
+            }
```
see above comment

andersonjackc replied:
Updated.

andersonjackc committed:
Added new IsEqual for admins to check equality of submission list

andersonjackc committed:
Update StandardUser IsEqualWithSubMap()

andersonjackc committed:
Updated 2 tests

Rather than breaking from the loops based on the counter getting too big, the loops do not run if they are not the same size, based on an additional test, per Josh's recommendation.

andersonjackc committed:
Updating the Submission/AnswerSubmission so that it actually works

andersonjackc committed:
Update DBConnector.cs

Now DBTests do not run if users are connected to the DB

andersonjackc committed:
Update Mailer.cs

Change to Mailer so that it can access info.txt in its new location

andersonjackc committed:
Update .gitignore

added info.txt to gitignore
